proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,160 @@
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not an int, uint, nor a floating-point type, or vectors 
+or matrices containing the aforementioned types, will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for TypedBuffers.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom builtins to use as type_traits, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every TypedBuffer resource type that is defined, so that all
+TypedBuffer HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are RawBuffer variants, such as `StructuredBuffer`. These
+resource types will have a different set of type-traits applied, which will
+loosen constraints on viable RETs.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. The 
+preprocessor will take responsibility for selecting the right set of type_traits
+to check, depending on the target IR. If DXIL is given, then all of the TypedBuffer
+`type_traits` will be applied on each TypedBuffer HLSL resource type. Otherwise, the
+RawBuffer type_traits will be applied to each resource type. For every `type_trait`
+that is not true for the given RET, an associated error message will be emitted.
+
+The list of type_traits that will be available for use are described below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_resource_element_type` | An RET should be an arithmetic type, or a bool, or a vector or matrix or UDT containing such types. |
+| `__builtin_is_homogenous` | A TypedBuffer RET with the DXIL IR target should never have two different subelement types. |
+| `__builtin_is_contained_in_four_groups_of_thirty_two_bits` | A TypedBuffer RET with the DXIL IR target should not have more than 4 elements, and the total size of the RET may not exceed 128 bits. |

```

Comment:
I feel like we could simplify this down to something like `__builtin_hlsl_is_line_vector_layout_compatible` to mark that a type is layout compatible with a [line-width](https://github.com/microsoft/hlsl-specs/pull/321/files#diff-29269dd46a5c0ae5cce91d3af91fce8d966c04887b38fe30ec54cb37cb28c5ceR300) vector. That would handle requiring homogeneousness, element count, and total size.

Reply:
Talked to Chris offline, we agreed `__builtin_hlsl_is_line_vector_layout_compatible` could be simplified by using the size_of builtin and other template tricks to verify vector length and total size in bytes.